presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/QueryCardinalityUtil.java

```diff
@@ -0,0 +1,142 @@
+    }
+
+    public static boolean isScalar(PlanNode node, Lookup lookup)
+    {
+        return Range.singleton(1L).encloses(extractCardinality(node, lookup));
+    }
+
+    public static boolean isAtMostScalar(PlanNode node)
+    {
+        return isAtMostScalar(node, noLookup());
+    }
+
+    public static boolean isAtMostScalar(PlanNode node, Lookup lookup)
+    {
+        return Range.closed(0L, 1L).encloses(extractCardinality(node, lookup));
+    }
+
+    private static Range<Long> extractCardinality(PlanNode node, Lookup lookup)
+    {
+        return node.accept(new CardinalityExtractorPlanVisitor(lookup), null);
+    }
+
+    private static final class CardinalityExtractorPlanVisitor
+            extends PlanVisitor<Range<Long>, Void>
+    {
+        private final Lookup lookup;
+
+        public CardinalityExtractorPlanVisitor(Lookup lookup)
+        {
+            this.lookup = requireNonNull(lookup, "lookup is null");
+        }
+
+        @Override
+        protected Range<Long> visitPlan(PlanNode node, Void context)
+        {
+            return Range.atLeast(0L);
+        }
+
+        @Override
+        public Range<Long> visitGroupReference(GroupReference node, Void context)
+        {
+            return lookup.resolve(node).accept(this, context);
+        }
+
+        @Override
+        public Range<Long> visitEnforceSingleRow(EnforceSingleRowNode node, Void context)
+        {
+            return Range.singleton(1L);
+        }
+
+        @Override
+        public Range<Long> visitAggregation(AggregationNode node, Void context)
+        {
+            if (node.hasEmptyGroupingSet()) {
+                return Range.singleton(1L);
+            }
+            return Range.atLeast(0L);
+        }
+
+        @Override
+        public Range<Long> visitExchange(ExchangeNode node, Void context)
+        {
+            if (node.getSources().size() == 1) {
+                return getOnlyElement(node.getSources()).accept(this, null);
+            }
+            return Range.atLeast(0L);
+        }
+
+        @Override
+        public Range<Long> visitProject(ProjectNode node, Void context)
+        {
+            return node.getSource().accept(this, null);
+        }
+
+        @Override
+        public Range<Long> visitFilter(FilterNode node, Void context)
+        {
+            Range<Long> sourceCardinalityRange = node.getSource().accept(this, null);
+            return Range.closed(0L, sourceCardinalityRange.upperEndpoint());
+        }
+
+        public Range<Long> visitValues(ValuesNode node, Void context)
+        {
+            return Range.singleton((long) node.getRows().size());
+        }
+
+        @Override
+        public Range<Long> visitLimit(LimitNode node, Void context)
+        {
+            Range<Long> sourceCardinalityRange = node.getSource().accept(this, null);
+            return Range.closed(sourceCardinalityRange.lowerEndpoint(), min(sourceCardinalityRange.upperEndpoint(), node.getCount()));
```

Comment:
`sourceCardinalityRange.lowerEndpoint()` might higher than `node.getCount()`, so range will be incorrect